### PR TITLE
Scripts to install k8s on ubuntu cluster with flannel network

### PR DESCRIPTION
**Pull request by resouer**
Some guys in Docker meetup said kubernetes is too heavy to use but we don't agree with that.

Kubernetes is simple and powerful, we can bootstrap a local cluster easily. A local cluster is much easier to enjoy, demo, hack, and even ship to production env with a little help from saltstack or puppet.

Our group is using these scripts and docs inspired by @jainvipin 's single node work to deploy kuber cluster. It works very well.

### CODE REVIEW

**Comment by tpires** [comment]
:+1: thanks!

**Comment by satnam6502** [comment]
@jbeda please can you review this or bounce it back to me with a suggestion for another reviewer? Thanks.

**Comment by zmerlynn** [comment]
I'm not against this pull, but see #3912 and suggested format #3926. @resouer, are you willing to own this going forward? (I'm also not the best person to review this, so not it.)

**Comment by jainvipin** [comment]
Good work @resouer 
This is a good way to demonstrate simplicity of kubernetes in multi-host runs. Ideally, instead of being specific to flannel (which is not a required component to run multi-host networking), I was hoping we also document  multi-host deployment as natively available in Kubernetes.

**Comment by tpires** [comment]
I've done a similar setup, regarding the init/init.d/default scripts, but I wanted them to start at boot. Based on [@pires](https://github.com/pires/kubernetes-vagrant-coreos-cluster/blob/master/master.yaml) I've done 2 waiter scripts:
- Same waiter script to wait for etcd to be available, on flanneld init;
- Other waiter script, to wait for /run/flannel/subnet.env file, for docker init.

**Comment by jbeda** [comment]
I'm sorry -- but I don't have time to review this before I disappear.  I need to get some stuff finished up.

**Comment by satnam6502** [comment]
@zmerlynn : can you take this or suggest someone else?

**Comment by jainvipin** [comment]
@zmerlynn, @satnam6502 please let me know if you'd like me to review these changes.

**Comment by zmerlynn** [comment]
@jainvipin: Please do, as it's based on your work. And between you and @resouer, I'm looking for an owner once this Getting Started Guide goes into the tree, because it'll need continued testing.

**Comment by WIZARD-CXY** [comment]
Although there was already a saltstack based solution, I think our solution is simpler and more agile, just a little more script. And I tested it twice in our lab. It all worked well.

I tried to build a k8s manually before, but only to find it failed due to a lack of overlay network support,as https://github.com/GoogleCloudPlatform/kubernetes/issues/2679.
@jainvipin As the brendandburns suggested ,I used flannel. The flannel is now simplest to provide a overlay network for k8s as far as I know.You just need a flanneld binary and a etcd network configure  key.And the performance of coreos/flannel is fine comparing to ovs solution I think.

**Comment by resouer** [comment]
@zmerlynn Yes. We are willing to own this work going forward and keeping updating it.  plz make that happen.

**Comment by jainvipin** [comment]
@satnam6502 @zmerlynn @resouer -  I am trying the changes, while I also review the code. Will get back on the thread with my comments and suggestions by today

**Comment by jainvipin** [comment]
@resouer @WIZARD-CXY, some comments on the changes:

#### Doc comments:
- /opt/bin should be in the $PATH to run etdctl as specified in the step 2, etc.
- Minor: You probably want to to make the cluster not specific to '4' nodes. I tried out with 2 nodes below (one master and one minion)

#### Some initial feedback from a quick try out:

I used the following etcd version
```
# etcd --version
etcd version 2.0.0
```

[1]
etcd launch failure on master: says etcd spawining too fast 

From /var/log/syslog it seems 'etcd main proess returned error 1'. The reason for etcd failure from /var/log/upstart/etcd.log is
```
2015/01/31 07:51:57 etcd: couldn't find local name "172.29.205.104" in the initial cluster configuration
```

Following the instructions from [Etcd Clustering](https://github.com/coreos/etcd/blob/master/Documentation/clustering.md), it seems like we should have used clustering configuration here.

[2]
etcd launch failure on minion: it seems like potions -peers isn't supported 
anymore, Either you require using etcd 1.5 version or with 2.0 change the 
scripts to use to appropriate options to specfi the peer/cluster URLs

Let me know if I should try this with 1.5 etcd, or more importantly since etcd2.0 is available now why not make it work with that version using the clustering (including discovery, etc. is very powerful)

I will give more feedback on the upstart scripts, etc. after above is resolved.

**Comment by jainvipin** [comment]
@WIZARD-CXY precisely my point too as discussed in #2679 that these steps are needed if you need virtualization layer on top of kubernetes networking model. Where as if you don't these do not apply - something yet to have a 'getting started guide' for. 
My suggestion would be to keep everything as is, and suggest the applicability to the deployment use case explicitly. 

> 
I tried to build a k8s manually before, but only to find it failed due to a lack of overlay network support,as #2679, brendandburns suggested I used flannel,so I used it and succeeded with little knowledge of it.
The reason we use flannel is because @jainvipin

**Comment by resouer** [comment]
@jainvipin  Thanks for your suggesting. We got your point. 

I'll update the document for (1) not specific for 4 nodes (2) k8s cluster can be built up natively without help of flannel and why we should do that in this case (3) the $PATH will be updated in the scripts you do not need to do it manually

On the other hand, we are using a etcd binary with very low version, sorry for not updating it, as etcd 2.0 is stable now, we'll update etcd configuration and specify all the dependencies's recommended version in the doc.

**Comment by jainvipin** [comment]
@resouer - regarding your comment (3) above regarding $PATH getting updated within upstart script, this takes care of launching etcd  when run using upstart scripts; however, it does not take care of running `etcdctl` which is recommended in step3 in readme.md file, e.g.

>  On any node:
>
> `$ etcdctl mk /coreos.com/network/config '{"Network":"10.0.0.0/16"}'`

I was suggesting adding /opt/bin to $PATH or change the above command to `$ /opt/bin/etcdctl mk /coreos.com/network/config ...`

**Comment by WIZARD-CXY** [comment]
We use 0.4.6 stable version of etcd in this guide. As etcd recently updated from 0.4.6 to 2.0, I think there must be a great number of people using version 0.4.6 over 2.0 . Although etcd version is not very important to the k8s ,I will update the configuration of 2.0 asap.And I will take your suggestion of changing the "etcdctl ..." command to $ /opt/bin/etcdctl mk /coreos.com/network/config ... @jainvipin

**Comment by WIZARD-CXY** [comment]
@jainvipin As the doc said the minion number can be customize by the user. And we think 3 minions choice is a reasonable one . 1 master ,1 minion k8s is a little small, and can only run 1 replica of pod which is not enough for even a guest example of k8s .

**Comment by resouer** [comment]
@jainvipin @WIZARD-CXY & me have updated the scripts and docs, please go on!

**Comment by WIZARD-CXY** [comment]
It was tested on 1 master node , 3 minion node successfully twice.In fact the 2.0.0 etcd configuration is a little tedious I think.

**Comment by jainvipin** [comment]
@resouer @WIZARD-CXY - Thanks for updating the scripts/docs!

Hmm... etcd 2.0 is IMO less tedious for some serious deployments with cluster discovery, etc. even though initially it might look a bit cumbersome. Anyways, I will get back by today after trying few things out...

**Comment by WIZARD-CXY** [comment]
I agree with you @jainvipin. But minion nodes may be in the private network. They can't easily access the Public etcd Discovery Service,so we use static etcd clustering approach instead although it is cumbersome to configure first time.

**Comment by jainvipin** [comment]
@WIZARD-CXY @resouer - just got around to looking at this today. Some comments:

##### Configure.sh script asking too much information
(Sorry for being picky here but if I were to do it I would take care of these)
- It certainly can avoid asking etcd version, which can be detected using `etcd --version` without having to run etcd
- It can avoid asking minion's addresses multiple times, once for etcd as URL list, another time for something else (some kube services I think)
- Entering URLs is not needed if they can be prepended and constructed in the shell scripts. You can also avoid asking the port#s and pick one for the user (2380 or 7001 whatever)

As a result ubuntu_multinodes_cluster.md is verbose and filled with if/else clauses, which it should not have been.

##### Ability to run minon on the master node as well
- When I enter master node as one of the minion address, it doesn't work for me. This is a result of how configure.sh handles populating /etc/default/kubelet, and /etc/default/kube-controller-manager, perhaps because the script assumes the node is either a minion or a master, but not both. So in my case, /etc/default/kublet is populated incorrectly as:
```--hostname_override=MY_IP \```
And /etc/default/kube-controller-manager only has one minion IP, instead of two comma separated ones.
It is good to allow this but if you don't want to in your instructions, perhaps a clarification in the docs would help. Or the output of `kubectl get minions` would help understand that too.

##### Starting and stopping can be made easier (this is a nice to have)
If you chain the dependencies (see single node example) to kick-start other services automatically then starting and stopping service can be easier. But I will leave this to you.

Appreciate your patience!

**Comment by resouer** [comment]
@jainvipin Thanks for your suggestions. We are re-factoring the workflow to alleviate burdens on users. This may included a bit more effort, so please stay tuned.

**Comment by jainvipin** [comment]
@resouer - sure, thanks for incorporating the suggestions.

@WIZARD-CXY regarding the following comment, I think a two node cluster is reasonable tryout if the deployment allows running minion on the node which is also running the master. Clearly, I can run it with some tweaks to /etc/default/kube-controller-server and /etc/default/kubelet. i.e .if you help incorporate the comment mentioned above, that would take care of things.

> @jainvipin As the doc said the minion number can be customize by the user. And we think 3 minions choice is a reasonable one . 1 master ,1 minion k8s is a little small, and can only run 1 replica of pod which is not enough for even a guest example of k8s which need 3 frontend pod .

**Comment by WIZARD-CXY** [comment]
@jainvipin I will take your good advise and modify the script to make it more easy and robust to use

**Comment by resouer** [comment]
@jainvipin We have updated the shell. Please review it.

**Comment by resouer** [comment]
@jainvipin We have updated the shell. Please review it.

**Comment by resouer** [comment]
@jainvipin We have updated the shell. Please review it. Thanks a lot!

**Comment by WIZARD-CXY** [comment]
We have updated the script and guide to meet all your requirement！ NOW it
is easier to use and have tested in our lab .please review it and see if it works for you@jainvipin

**Comment by jainvipin** [comment]
@WIZARD-CXY @resouer 
I tried out the recent changes, looks much cleaner and better - thanks for incorporating the suggestions!

I see only one issue with the script (rest all looks good):

```
$ sudo ./configure.sh 
Welcome to use this script to configure k8s setup

Please enter all your cluster node ips, MASTER node comes first
And separated with blank space like "<ip_1> <ip_2> <ip_3>": 172.29.205.104 172.29.205.105

This machine acts as
  both MASTER and MINION:      1
  only MASTER:                 2
  only MINION:                 3
Please choose a role > 1

IP address of this machine > 172.29.205.104

Configure Success

$ cat /etc/default/kube-controller-manager | grep machines
--machines=172.29.205.105 \
```

The machines configuration in kube-controller-manager should have been both master and minion. Upon debugging it seems  like you got the options flipped in your configuration. Following patch seems 
to take care of the issue:

```
diff --git a/cluster/ubuntu-cluster/configure.sh b/cluster/ubuntu-cluster/configure.sh
index 3f22471..99ef687 100755
--- a/cluster/ubuntu-cluster/configure.sh
+++ b/cluster/ubuntu-cluster/configure.sh
@@ -153,7 +153,7 @@ while true; do
     echo
 
        case $option in
-           [1] )
+           [3] )
                read -p "IP address of this machine > " myIP
             echo
             etcdName=${mm[$myIP]}
@@ -175,7 +175,7 @@ while true; do
                cpMinion
                break
                ;;
-           [3] )
+           [1] )
                read -p "IP address of this machine > " myIP
             echo 
             etcdName=${mm[$myIP]}
@@ -198,4 +198,4 @@ while true; do
        esac
 done
 
-echo -e "\e[0;32mConfigure Success\033[0m"
\ No newline at end of file
+echo -e "\e[0;32mConfigure Success\033[0m"
```

Rest looks good, including docs.

**Comment by WIZARD-CXY** [comment]
Thanks for the bug report, and I have fixed it already.A little different from your patch.Please review it @jainvipin @resouer

**Comment by jainvipin** [comment]
LGTM!

**Comment by resouer** [comment]
ping @zmerlynn

**Comment by WIZARD-CXY** [comment]
ok,CLA is good

**Comment by zmerlynn** [comment]
Looking now, was OOO.

File: cluster/ubuntu-cluster/build.sh [hunk 1]
```diff
@@ -0,0 +1,66 @@
+#!/bin/bash
+
+# Copyright 2014 Google Inc. All rights reserved.
```
Comment (zmerlynn) [inline cluster/ubuntu-cluster/build.sh]: It's 2015

**Comment by zmerlynn** [comment]
Please fix the copyrights on new files. Is there any way to squash these commits? 23 is pretty heavy.

**Comment by WIZARD-CXY** [comment]
OK I will fix this asap

**Comment by WIZARD-CXY** [comment]
@zmerlynn just squash commits to 2 commits and fix the year issue on copyrights.

**Comment by brendandburns** [comment]
Will merge tomorrow during daytime merge hours.

**Comment by WIZARD-CXY** [comment]
OK

**Pull request merged**